content/blog/2024-07-10-gsoc-multiboot2.mdx

```diff
@@ -0,0 +1,106 @@
+
+## Testing the Waters: Progress and Challenges
+
+My initial focus was to lay the groundwork by creating the `multiboot.S` and `multiboot2.py` files and testing them with the simple `helloworld` application.
+This allowed me to verify that the Multiboot2 header was inserted correctly and the system could boot without major hiccups.
+
+Of course, debugging became my constant companion during this process.
+Tools like `hexdump` and attaching `gdb` to my custom GRUB build proved invaluable.
+
+I encountered an initial hurdle regarding the ELF file.
+After playing around with GRUB error messages and making a "comparative study" of the binaries with Multiboot, Multiboot2, or no additional header, I realized the ELF was lacking the Multiboot2 header altogether.
+Some detective work later, I traced the issue to an unlinked `multiboot2.py` script.
+
+## Multiboot2 Header: A Deep Dive
+
+So far, I really emphasized the need of the Multiboot2 header.
+But what is its role?
+
+The Multiboot2 header is a structured data block embedded within the bootable system image.
+Its purpose is to facilitate the transition of control and information from the bootloader to the kernel during the boot process.
+This header acts as a standardized communication channel, ensuring the kernel can access essential details about the system's hardware, memory layout, and other boot-related parameters.
+
+### Format and Contents
+
+The Multiboot2 header follows a specific format, starting with a fixed-size header followed by a series of variable-length tags.
+
+The fixed-size header includes:
+
+- The magic number `0xE85250D6` indicating a Multiboot2-compliant image
+- The target architecture  (e.g. `i386`, `x86_64`)
+- The header length, including all tags
+- The checksum of the header, used for error detection
+
+I talked a little about the tags in my previous blog post, but to recap, they provide detailed information about the system's memory layout, boot command line, and other essential parameters.
+Commonly, each tag has a type, size and payload.
+There is a multitude of tags to choose from, depending on the specific requirements of the system, adding to the flexibility of Multiboot2.
+
+Additionaly, in the context of this project, it may be valuable to note that simply adding the Multiboot2 header to the ELF file is not sufficient.
+On top of that, both `multiboot.py` and `multiboot2.py` scripts add an extra ELF header, "sandwiching" the Multiboot2 header between it and the original ELF.
+This ensures that GRUB identifies the overall file as an ELF and avoids any potential complications along the way.
+
+### Interaction between Bootloader and Kernel
+
+The bootloader locates the Multiboot2 header within the ELF file.
+After verifying the magic number and architecture compatibility, it parses the header and tags to gather essential system information.
+It then loads the kernel into memory, sets up the environment, and transfers control to the kernel's entry point.
```

Comment:
I would say here that the bootloader, after finding the information requested by the tags in the header, places that information in the kernel memory, then passes the control, for clarity